### PR TITLE
8319429: Resetting MXCSR flags degrades ecore

### PR DESCRIPTION
**Pull request by vpaprotsk**
Improves vector rounding on ECore about 10x
```
(BEFORE) FpRoundingBenchmark.test_round_float        2048  thrpt    3  40.912 ± 0.044  ops/ms
(AFTER ) FpRoundingBenchmark.test_round_float        2048  thrpt    3  431.682 ± 0.727  ops/ms
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319429](https://bugs.openjdk.org/browse/JDK-8319429): Resetting MXCSR flags degrades ecore (**Enhancement** - P4)

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16504/head:pull/16504` \
`$ git checkout pull/16504`

Update a local copy of the PR: \
`$ git checkout pull/16504` \
`$ git pull https://git.openjdk.org/jdk.git pull/16504/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16504`

View PR using the GUI difftool: \
`$ git pr show -t 16504`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16504.diff">https://git.openjdk.org/jdk/pull/16504.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16504#issuecomment-1793207076)

### CODE REVIEW

File: src/hotspot/share/runtime/globals.hpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/globals.hpp:574]: I think this should be a CPU specific flag in ./cpu/x86/globals_x86.hpp (similar to how we have linux specific flags in ./os/linux/globals_linux.hpp). Also the description should clarify that the default is actually true for Ecore systems, and false elsewhere.

File: src/hotspot/cpu/x86/vm_version_x86.cpp
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:864]: And what should happen if the flag is set true by the user and there is no Ecore? What affect will that have? Should it be allowed?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/globals.hpp
Comment (vpaprotsk) [reply src/hotspot/share/runtime/globals.hpp:574]: Thanks, didn't know about that file, will try.

File: src/hotspot/cpu/x86/vm_version_x86.cpp
Comment (vpaprotsk) [reply src/hotspot/cpu/x86/vm_version_x86.cpp:864]: From ISA point of view, they are the same, so if user flips flag on purpose, code will still be correct. Its also helpful to test Ecore optimized code on a Pcore (I have some more patches coming in under this option soon)

File: src/hotspot/cpu/x86/vm_version_x86.cpp
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:862]: Model 0x97 corresponds to ADL Hybrid Core converged ISA target where DoEcodeOpt enabling will depend on the core over which VM initialization thread executes.

Comment (vpaprotsk) [reply src/hotspot/cpu/x86/vm_version_x86.cpp:862]: On hybrid platforms, it will be enabled on both Pcore and Ecore. Performance is essentially unchanged on Pcore

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/vm_version_x86.cpp:862]: Correct, CPUID is CPU specific not core specific, so settings should apply to both.

File: src/hotspot/cpu/x86/x86.ad
Comment (vpaprotsk) [reply src/hotspot/cpu/x86/x86.ad:7443]: Not convinced that it makes it cleaner; while reading, means one extra file lookup to see what the macro does, vs how it immediately is clear now (ctags working half the time..). Macros seem to be paired better here with conditional compilation, don't see them paired with dev options.. 

I can perhaps be convinced, it _is_ repeated 6 times in this PR. Perhaps `globals_x86.hpp` might be an acceptable place, but doesn't appear to have a precedent there.

File: src/hotspot/cpu/x86/vm_version_x86.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/vm_version_x86.cpp:862]: As per JVM specification section 2.8 "The floating-point instructions of the Java Virtual Machine do not throw
exceptions, trap, or otherwise signal the IEEE 754 exceptional conditions of
invalid operation, division by zero, overflow, underflow, or inexact.", thus JVM does not check these exceptions.

Your patch is always setting lower 6 bits of MXCSR on hybrid CPUs which has both E and P core, do you see any concerns if these bits are default ON for other server targets.

**Comment by jatin-bhateja** [comment]
As per JVM specification section 2.8 "The floating-point instructions of the Java Virtual Machine do not throw
exceptions, trap, or otherwise signal the IEEE 754 exceptional conditions of
invalid operation, division by zero, overflow, underflow, or inexact.", thus JVM does not check these exceptions.

Your patch is always setting lower 6 bits of MXCSR on hybrid CPUs which has both E and P core, do you see any concerns if these bits are default ON for other server targets.

**Comment by vpaprotsk** [comment]
> As per JVM specification section 2.8 "The floating-point instructions of the Java Virtual Machine do not throw exceptions, trap, or otherwise signal the IEEE 754 exceptional conditions of invalid operation, division by zero, overflow, underflow, or inexact.", thus JVM does not check these exceptions.
> 
> Your patch is always setting lower 6 bits of MXCSR on hybrid CPUs which has both E and P core, do you see any concerns if these bits are default ON for other server targets with just P-cores.

I considered it. There doesn't appear to be any functional correctness issues, since Java does not support the signaling part of BFP IEEE anyway, those flags are essentially noop. I also measured on a some PCore systems, the performance is unaffected.  I mostly went with this fix to be conservative, since there probably should be more performance testing otherwise. Might be cleaner to have it set to just one value.

**Comment by jatin-bhateja** [comment]
> > As per JVM specification section 2.8 "The floating-point instructions of the Java Virtual Machine do not throw exceptions, trap, or otherwise signal the IEEE 754 exceptional conditions of invalid operation, division by zero, overflow, underflow, or inexact.", thus JVM does not check these exceptions.
> > Your patch is always setting lower 6 bits of MXCSR on hybrid CPUs which has both E and P core, do you see any concerns if these bits are default ON for other server targets with just P-cores.
> 
> I considered it. There doesn't appear to be any functional correctness issues, since Java does not support the signaling part of BFP IEEE anyway, those flags are essentially noop. I also measured on a some PCore systems, the performance is unaffected. I mostly went with this fix to be conservative, since there probably should be more performance testing otherwise. Might be cleaner to have it set to just one value.

Do you have any idea why these settings give a performance bump over E-core, are these suggested settings in x86 manuals?

File: src/hotspot/cpu/x86/globals_x86.hpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/globals_x86.hpp:218]: Comment addressal is pending.

**Comment by vpaprotsk** [comment]
@dholmes-ora did you have any further concerns? Would you mind running any extra tests you have?

**Comment by dholmes-ora** [comment]
> @dholmes-ora did you have any further concerns? Would you mind running any extra tests you have?

I have no further concerns - thanks for moving the flag. Someone from our compiler team should review this and run it through our CI. I don't know if we have any machines that will be affected by this.

**Comment by sviswa7** [comment]
@TobiHartmann @vnkozlov Could we please get one more review on this small PR?

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me. I'll run it through our testing and report back.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by sviswa7** [comment]
Thanks a lot Tobias!

**Pull request closed without merging**
